### PR TITLE
Fix hashring for all possible failure scenarios

### PR DESCRIPTION
**Pull request by spaparaju**
This PR adds a new flag 'allow-only-ready-replicas' for hashring to contain Thanos Receive replicas which are in the 'Ready' status.

Under this flag, this PR includes fixes for : 
1. hashring containing non-Ready replicas if a replica turns from 'Ready' status to  a 'non-ready' status.
2. hashring containing non-Ready replicas if a scale up does not succeed to the completion (spec.Replicas != status.ReadyReplicas)

Signed-off-by: SriKrishna Paparaju <paparaju@gmail.com>

### CODE REVIEW

File: main.go
Comment (ianbillett) [inline main.go:573]: I don't think that this section is doing what we think it is doing here.

[c.waitForPod](https://github.com/spaparaju/thanos-receive-controller/blob/b3973a78730ff0dea1c48e699787b4de6e4e20a6/main.go#L545) waits until the pod reaches `Running` [phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase), meaning the containers have all started. 

A pod is `Ready` when the application passes the configured `readinessProbe` indicating that it is ready to serve traffic.

**Review by ianbillett — COMMENTED** [review]
Reading the code it appears that we have always mixed up a pod in `Running` phase with a container `Ready` to serve requests - it would be good to clean that up in this PR.

Also - I'm wondering if there is a smarter way we can be informed of changes to the hashrings we are watching? 

The [newFilteredStatefulSetInformer](https://github.com/spaparaju/thanos-receive-controller/blob/b3973a78730ff0dea1c48e699787b4de6e4e20a6/main.go#L342-L344) function accepts a filtering function - perhaps we could parse out the pods that are suitable there?

**Comment by spaparaju** [comment]
> Reading the code it appears that we have always mixed up a pod in `Running` phase with a container `Ready` to serve requests - it would be good to clean that up in this PR.
> 
> Also - I'm wondering if there is a smarter way we can be informed of changes to the hashrings we are watching?
> 
> The [newFilteredStatefulSetInformer](https://github.com/spaparaju/thanos-receive-controller/blob/b3973a78730ff0dea1c48e699787b4de6e4e20a6/main.go#L342-L344) function accepts a filtering function - perhaps we could parse out the pods that are suitable there?
> 
> edit: I think we could use the [ListOptions.FieldSelectors](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ListOptions) to do this.

pods get to 'Running' status once the 'readiness' probe is successful.
The current [sync loop](https://github.com/observatorium/thanos-receive-controller/blob/cd12a419fc00eeb354d1f48e3077648222b16c07/main.go#L474) leverages status changes notified through Kube. informers.

File: main.go
Comment (spaparaju) [reply main.go:573]: replicas get to 'Running' status once their readiness probe is successful. Yes, there is [NO pod status](c.options.allowOnlyReadyReplicas) called 'Ready'. Are you referring to change the name of the variable 'c.options.allowOnlyReadyReplicas' to 'c.options.allowOnlyRunningReplicas' ?

Comment (matej-g) [reply main.go:573]: I think @ianbillett is indeed referring to the difference between pod's [_phase_](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase) and [_conditions_](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions). From my understanding, as @ianbillett said, a pod is considered to be running when all containers are started - but this does not necessarily mean the pod has passed the readiness probe and that it is capable of receiving requests. I'm therefore also wondering if trying to rely on this _condition_ as opposed to _phase_ would make more sense :thinking:.

Comment (spaparaju) [reply main.go:573]: In general, `readiness-probe` is meant for internal, pod life-cycling event triggering  changes to `pod status`. 
If `readiness-probe` is defined, until the `readiness-probe` is evaluated no traffic is sent to the respective pod. Failure of `readiness-probe` results in removing pod IP from all the referencing services so that the [failed pod does not get any traffic](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#:~:text=readinessProbe,the%20initial%20delay%20is%20Failure%20.) and triggers pod status change. 
`readiness-probe` is triggered all through lifetime of a pod , so at any point of time the readiness-probe failures result in pod status change.
I checked the [readiness probe](https://github.com/thanos-io/thanos/blob/fa4a40d5218709dc25c7be86af26100bab0516a6/pkg/receive/handler.go#L190) for `Thanos receive` and its not heavy (like loading cache etc.).

**Comment by ianbillett** [comment]
@spaparaju and I chatted about this PR last Thursday, there are a couple of things to do before merging:
* Test out this change in a local cluster and verify that we see the behaviour we desire.
* Eiher update the naming of the flag to be `running` rather than `ready` to better represent the work currently being done or update the mechanism for waiting to poll for `readiness` in the pods.

**Comment by spaparaju** [comment]
> @spaparaju and I chatted about this PR last Thursday, there are a couple of things to do before merging:
> 
> * Test out this change in a local cluster and verify that we see the behaviour we desire & report back results in this PR.
Tested on Minikube. Here are the recordings of the testing: [reproduce ](https://drive.google.com/file/d/1aT-aMUeyHzyJ2wuhKNquhTwlBXysIyE-/view?usp=sharing
)the hash-ring update problem and [how this PR](https://drive.google.com/file/d/1L1lZr5q79gXJzn6m_BfoFnA6FVhHvpc-/view?usp=sharing) solves the update problem. 
> * Eiher update the naming of the flag to be `running` rather than `ready` to better represent the work currently being done or update the mechanism for waiting to poll for `readiness` in the pods.
renamed the flag to include 'running' instead of  the word 'ready'

**Comment by ianbillett** [comment]
Really nice demos there @spaparaju what an effective way to do code review :) LGTM

**Review by ianbillett — APPROVED** [review]

**Review by matej-g — APPROVED** [review]
Looking good!

**Comment by matej-g** [comment]
The CI seem to be stuck, probably for some time already :disappointed:, we should take a look.

**Comment by ianbillett** [comment]
@bwplotka do you have perms to nuke this and start it again?

**Comment by matej-g** [comment]
Hm so I also don't seem to have permission to change settings, since I don't see the `Settings` tab for this repo, we might need to escalate to someone who is able to check if there is any related CI setting.

image

**Review by bwplotka — APPROVED** [review]
Not sure about change, but does not looks significant, lgtm

**Comment by bwplotka** [comment]
Unfortunately I don't have perms to this - also it looks like there is no link to drone CI so connection did not even start. Maybe @squat has some ideas?

If not we should spend 30m to move to githubActions quickly. Otherwsie we would merge this PR without any CI.

**Comment by squat** [comment]
I think that the hosted version of Drone CI hasn't been working so well ever since it was acquired by Harness or it has been entirely decommissioned :/ you can't even find https://cloud.drone.io on google anymore. I just disabled and re-enabled the project in Drone and now webhooks are working again, but CI still doesn't run. I think it's time to finally move this project onto GitHub Actions. And also rename master -> main. This repo needs a bit of TLC

**Comment by michael-burt** [comment]
I opened #89 to address the issue where non-ready replicas were being populated in the hashring configuration. My approach is similar to the one taken in this PR, however I filter only for `Ready` replicas since `Running` is not sufficient to ensure that the replica is able to accept write remote write requests.

**Comment by matej-g** [comment]
Unless there are objections, I think we should close this in favor of https://github.com/observatorium/thanos-receive-controller/pull/89, I believe the approach to change hashring only when scaling the stateful set to be preferable here.

**Pull request closed without merging**
